substrate-archive-backend/src/database.rs

```diff
@@ -75,33 +77,68 @@ impl ReadOnlyDatabase {
             }
         }
     }
+}
+
+impl ReadOnlyDB for SecondaryRocksDB {
+    fn get(&self, col: ColumnId, key: &[u8]) -> Option<Vec<u8>> {
+        self.get(col, key)
+    }
 
-    pub fn iter<'a>(&'a self, col: u32) -> impl Iterator<Item = KeyValuePair> + 'a {
-        self.inner.iter(col)
+    fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = KeyValuePair> + 'a> {
+        Box::new(self.inner.iter(col))
     }
 
-    pub fn try_catch_up_with_primary(&self) -> Result<()> {
+    fn catch_up_with_primary(&self) -> Result<()> {
         if self.track_catchups {
             self.catch_counter.fetch_add(1, Ordering::Relaxed);
         }
         self.inner.try_catch_up_with_primary()?;
         Ok(())
     }
 
-    pub fn catch_up_count(&self) -> Option<usize> {
-        if !self.track_catchups {
-            log::warn!("catchup tracking is not enabled");
-            None
-        } else {
-            Some(self.catch_counter.fetch_add(0, Ordering::Relaxed))
+    fn open_database(
+        path: &str,
+        cache_size: usize,
+        db_path: PathBuf,
+    ) -> sp_blockchain::Result<SecondaryRocksDB> {
+        // need to make sure this is `Some` to open secondary instance
+        let db_path = db_path.as_path().to_str().expect("Creating db path failed");
```

Comment:
Can this be something like?

```suggestion
        let db_path = db_path.as_path().to_str().map_err(|e| Error::Msg(format!("Bad database path: {:?}", e)?;
```

Reply:
#122